File: dogfight/src/entities/Bomb.ts
import { Entity, EntityType } from "../entity";
import { Cache, CacheEntry } from "../network/cache";
import { Team, SCALE_FACTOR } from "../constants";
import { radiansToDirection, directionToRadians } from "../physics/helpers";
import { RectangleBody, Rectangle } from "../physics/rectangle";
import { GameWorld, } from "../world/world";
import { PlayerInfo } from "./PlayerInfo";
import { SolidEntity } from "./SolidEntity";
import { Ownable } from "../ownable";
import { BufferedImage } from "../BufferedImage";
import { Runway } from "./Runway";
import { Plane } from "./Plane";
import { OwnableSolidEntity } from "./OwnableSolidEntity";

export const bombGlobals = {
  gravity: 425,
  drag: 0.005,
  cooldown: 500
};

export class Bomb extends OwnableSolidEntity {

  public type = EntityType.Bomb;
  public age: number;
  public localX: number;
  public localY: number;
  public x: number;
  public y: number;
  public droppedBy: number; // ID of player who dropped it
  public team: Team; // team of player who dropped it

  public image;
  public width;
  public height;

  public xSpeed: number;
  public ySpeed: number;
  public radians: number;
  public origin: OwnableSolidEntity;
  public filename = "bomb.gif";
  //public direction: number;

  public constructor(id: number, world: GameWorld, cache: Cache, x: number, y: number, direction: number, speed: number, origin: OwnableSolidEntity) {
    super(id, world, origin.getTeam());
    //this.radians = directionToRadians(direction);
    this.setDirection(cache, direction);
    this.setPos(cache, x, y);
    //this.localX = x * SCALE_FACTOR;
    //this.localY = y * SCALE_FACTOR;
    this.setSpeed(cache, speed * SCALE_FACTOR); // needs radians to be set
    this.image = world.getImage(this.filename);
    this.width = this.image.width;
    this.height = this.image.height;
    this.origin = origin;


    this.setData(cache, {
      age: 0,
      x: x,
      y: y,
      direction: direction,
      droppedBy: origin.getPlayerInfo().id,
      team: origin.getTeam()
    });
  }
  public getPlayerInfo(): PlayerInfo {
    return this.origin.getPlayerInfo();
    //throw new Error("Method not implemented.");
  }
  public getRootOwner(): Ownable {
    return this.origin.getRootOwner();
    //throw new Error("Method not implemented.");
  }
  public getCollisionBounds(): import("../physics/rectangle").Rectangle {
    //throw new Error("Method not implemented.");
    return new Rectangle(this.x, this.y, this.width, this.height);
  }
  public getCollisionImage(): BufferedImage {
    return this.world.getImage(this.filename + "_rot_" + radiansToDirection(this.radians) + "_flip_" + false);
  }
  public tick(cache: Cache, deltaTime: number): void {
    this.move(cache, deltaTime);
    this.ageBomb(cache, deltaTime);
    this.checkCollision();
  }

  private ageBomb(cache: Cache, deltaTime: number): void {
    this.age += deltaTime;
  }

  public hit(se: SolidEntity): void {
    console.log("bomb - hit");
    if (se == null) {
      this.getPlayerInfo().submitBomb(this, false, false);
      this.world.removeEntity(this);
    }
    else if (this.origin == null || !(this.origin === se)) { //&& !this.isRemoved()) 
      if (se.getType() == EntityType.ControlTower || se.getType() == EntityType.Runway) {
        if (se.getTeam() == this.getTeam()) {
          this.getPlayerInfo().submitTeamBomb(this, true);
        }
        else {
          this.getPlayerInfo().submitBomb(this, true, true);
        }
      }
      else if (se instanceof Plane) {
        if (se.getTeam() == this.getTeam()) {
          this.getPlayerInfo().submitTeamBomb(this, false);
        }
        else {
          this.getPlayerInfo().submitBomb(this, true, false);
        }
      }
      else {
        this.getPlayerInfo().submitBomb(this, false, false);
      }
      this.world.removeEntity(this);
      if (se.getType() != EntityType.Water) {
        this.world.createExplosion(this.x, this.y, this);
      }
    }

  }

  public move(cache: Cache, deltaTime: number): void {
    // move the bomb...
    //console.log(this.x, this.y, this.xSpeed);
    const tstep = deltaTime / 1000;
    const xDelta = tstep * 1; // 0.01 * 100 tps in original
    const yDelta = tstep * 1000 / 3; // 3.33 * 100 tps in original

    this.localX += this.xSpeed * tstep * SCALE_FACTOR;
    this.localY += this.ySpeed * tstep * SCALE_FACTOR;
    this.xSpeed -= this.xSpeed * xDelta;
    this.ySpeed -= yDelta;
    this.radians = Math.atan2(this.ySpeed, this.xSpeed);

    this.setData(cache, {
      x: Math.round(this.localX / SCALE_FACTOR),
      y: Math.round(this.localY / SCALE_FACTOR),
      direction: radiansToDirection(this.radians)
    });
  }

  public setSpeed(cache: Cache, scaledSpeed: number): void {
    this.xSpeed =
      Math.cos(this.radians) * Math.round(scaledSpeed);
    this.ySpeed =
      Math.sin(this.radians) * Math.round(scaledSpeed);
  }

  public setDirection(cache: Cache, dir: number): void {
    this.radians = directionToRadians(dir);
    this.set(cache, "direction", dir);
  }

  public setPos(cache: Cache, x: number, y: number): void {
    // console.log(x, y);
    this.localX = x * SCALE_FACTOR;
    this.localY = y * SCALE_FACTOR;
    // console.log(this.localX / 100, this.localY / 100);
    this.setData(cache, { x, y });
    // console.log(this.x, this.y);
  }

  public getState(): CacheEntry {
    return {
      type: this.type,
      age: this.age,
      x: this.x,
      y: this.y,
      direction: radiansToDirection(this.radians)
    };
  }
}

export function getBombRect(
  x: number,
  y: number,
  direction: number,
): RectangleBody {
  return {
    // width: Math.round(planeData[type].width * 0.8),
    // height: Math.round(planeData[type].height * 0.8),
    // TODO get from image
    width: 7,
    height: 10,
    center: { x, y },
    direction,
  };
}

File: dogfight/src/entities/bomb.rs
use dogfight_macros::Networked;
use image::RgbaImage;

use crate::{
    collision::{BoundingBox, SolidEntity},
    images::{get_rotateable_image, rotate_image, BOMB},
    math::radians_to_direction,
    network::{property::Property, EntityProperties, NetworkedEntity},
    world::RESOLUTION,
};

use super::{container::PlayerId, entity::Entity, types::EntityType};

#[derive(Networked)]
pub struct Bomb {
    player_id: PlayerId,
    x: i32,
    y: i32,
    client_x: Property<i16>,
    client_y: Property<i16>,
    direction: Property<u8>,
    angle: f64,
    x_speed: f64,
    y_speed: f64,
    image: RgbaImage,
    rotated_image: RgbaImage,
    is_alive: bool,
}

impl Bomb {
    pub fn new(player_id: PlayerId, x: i16, y: i16, angle: f64, speed: f64) -> Bomb {
        let dir = radians_to_direction(angle);
        // log(format!("angle: {} dir: {}", angle, dir));
        Bomb {
            player_id,
            x: (x as i32) * RESOLUTION,
            y: (y as i32) * RESOLUTION,
            client_x: Property::new(x),
            client_y: Property::new(y),
            direction: Property::new(dir),
            angle: angle,
            x_speed: angle.cos() * speed * 100.0,
            y_speed: angle.sin() * speed * 100.0,

            image: get_rotateable_image(BOMB),
            rotated_image: get_rotateable_image(BOMB),
            is_alive: true,
        }
    }

    pub fn kill(&mut self) -> () {
        self.is_alive = true;
    }

    pub fn player_id(&self) -> PlayerId {
        self.player_id
    }

    pub fn set_position(&mut self, x: i16, y: i16) {
        self.x = x as i32 * RESOLUTION;
        self.y = y as i32 * RESOLUTION;
        self.client_x.set(x);
        self.client_y.set(y);
    }

    pub fn tick(&mut self) {
        self.x += ((RESOLUTION as f64) * self.x_speed / (RESOLUTION as f64)) as i32;
        self.y += ((RESOLUTION as f64) * self.y_speed / (RESOLUTION as f64)) as i32;

        self.client_x.set((self.x / RESOLUTION) as i16);
        self.client_y.set((self.y / RESOLUTION) as i16);

        self.x_speed -= self.x_speed * 0.01;
        self.y_speed += 3.3333333333333335;

        // update angle/direction
        self.angle = self.y_speed.atan2(self.x_speed);
        // For some reason the rust version is returning negative angles with atan2
        // and the java version doesn't, maybe the java version y coords are flipped?
        // Either way, ChatGPT says this is the way to flip it and it seems to work
        if self.angle < 0.0 {
            self.angle += std::f64::consts::TAU; // stop negative angle values
        }
        self.direction.set(radians_to_direction(self.angle));
    }

    pub fn get_x(&self) -> i16 {
        *self.client_x.get()
    }

    pub fn get_y(&self) -> i16 {
        *self.client_y.get()
    }
}

impl Entity for Bomb {
    fn get_type(&self) -> EntityType {
        EntityType::Bomb
    }
}

impl SolidEntity for Bomb {
    fn get_collision_bounds(&self) -> BoundingBox {
        BoundingBox {
            x: (self.x / RESOLUTION) as i16,
            y: (self.y / RESOLUTION) as i16,
            width: self.image.width() as i16,
            height: self.image.height() as i16,
        }
    }

    fn get_collision_image(&self) -> Option<&RgbaImage> {
        Some(&self.rotated_image)
    }

    fn do_rotate_collision_image(&mut self) -> () {
        self.rotated_image = rotate_image(&self.image, self.angle, false);
    }

    fn is_alive(&self) -> bool {
        self.is_alive
    }
}
